pkg/syncer/pvcsi_fullsync.go

```diff
@@ -233,6 +241,121 @@ func createCnsVolumeMetadataList(ctx context.Context, metadataSyncer *metadataSy
 	return nil
 }
 
+// AddNodeAffinityRulesOnPV update TKG PVs with node affinity rules set on associated supervisor PVC if
+// node affinity rule is not set on the TKG PV
+func AddNodeAffinityRulesOnPV(ctx context.Context, metadataSyncer *metadataSyncInformer) {
+	log := logger.GetLogger(ctx)
+	log.Info("AddNodeAffinityRulesOnPV Start.")
+	pvList, err := getPVsInBoundAvailableOrReleased(ctx, metadataSyncer)
+	if err != nil {
+		log.Errorf("FullSync: Failed to get PVs from guest cluster. Err: %v", err)
+		return
+	}
+	// Get the supervisor namespace in which the guest cluster is deployed.
+	supervisorNamespace, err := cnsconfig.GetSupervisorNamespace(ctx)
+	if err != nil {
+		log.Errorf("FullSync: could not get supervisor namespace in which guest cluster was deployed. Err: %v", err)
+		return
+	}
+
+	// Create the kubernetes client from config.
+	k8sClient, err := k8s.NewClient(ctx)
+	if err != nil {
+		log.Errorf("creating Kubernetes client failed. Err: %v", err)
+		return
+	}
+
+	for _, pv := range pvList {
+		if pv.Spec.NodeAffinity == nil {
+			supervisorPVClaim, err := metadataSyncer.supervisorClient.CoreV1().
+				PersistentVolumeClaims(supervisorNamespace).Get(ctx, pv.Spec.CSI.VolumeHandle, metav1.GetOptions{})
+			if err != nil {
+				log.Errorf("AddNodeAffinityRulesOnPV: failed to get supervisor PVC for the PV %v. Err: %v",
+					pv.Name, err)
+				continue
+			}
+			volumeAccessibleTopologyJSON := supervisorPVClaim.Annotations["csi.vsphere.volume-accessible-topology"]
+			var accessibleTopology []*csi.Topology
+			if volumeAccessibleTopologyJSON != "" {
+				accessibleTopology, err = parseJSONToTopologies(volumeAccessibleTopologyJSON)
+				if err != nil {
+					log.Errorf("failed to parse json %v to Topology. err :%v", volumeAccessibleTopologyJSON, err)
+					continue
+				}
+			}
+			pvnodeAffinityRules := GenerateVolumeNodeAffinity(accessibleTopology)
+			pv.Spec.NodeAffinity = pvnodeAffinityRules
+			_, err = k8sClient.CoreV1().PersistentVolumes().Update(ctx, pv, metav1.UpdateOptions{})
+			if err != nil {
+				log.Errorf("failed to update PV %v with node affinity rules %v. err :%v",
+					pv.Name, pvnodeAffinityRules, err)
+				continue
+			}
+			log.Infof("patched PV: %v with node affinity %v", pv.Name, pvnodeAffinityRules)
+		}
+	}
+	log.Info("AddNodeAffinityRulesOnPV End.")
+}
+
+// Convert JSON array string to a slice of Topology objects
+func parseJSONToTopologies(jsonStr string) ([]*csi.Topology, error) {
```

Comment:
Why do we need to do all this? Can't we use a simpler logic like the one used in https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/ba87988dcc48aa49a06dacfa6c3adeeb4c4e1a4e/pkg/csi/service/wcpguest/controller_helper.go#L283?

Reply:
yes this is better. I have replaced it with that function